pandas/core/indexes/datetimelike.py

```diff
@@ -547,31 +573,103 @@ def _maybe_box_as_values(self, values, **attribs):
         # - sort_values
         return values
 
+    def _deepcopy_if_needed(self, orig, copy=False):
```

Comment:
woa? what the heck is this?

Reply:
Without this, tests like `pandas/tests/indexes/datetimes/test_astype.py::TestDatetimeIndex::test_integer_index_astype_datetime[US/Pacific-datetime64[ns, US/Pacific]]` fail.

`Index._deepcopy_if_needed` seems to assume that `._data` is an ndarary. It accesses attributes like `self._data.base`, which doesn't exist on a DTA.

We could potentially add a `_ref_to_data` attribute, which for DatetimeLikeIndexes would be `_data._data`?

Reply:
what calls this? this seems like a work-around for an actual bug

Reply:
Seems like DatetimeIndex.__new__ calls it on `ref_to_data`

```python
In [9]: idx = pd.date_range('2000', periods=4)

In [10]: pd.DatetimeIndex(idx._data)
> /Users/taugspurger/sandbox/pandas/pandas/core/indexes/datetimelike.py(577)_deepcopy_if_needed()
-> if copy:
(Pdb) up
> /Users/taugspurger/sandbox/pandas/pandas/core/indexes/datetimes.py(291)__new__()
-> return subarr._deepcopy_if_needed(ref_to_data, copy)
(Pdb) l
286                 dayfirst=dayfirst, yearfirst=yearfirst, ambiguous=ambiguous)
287
288             subarr = cls._simple_new(dtarr._data, name=name,
289                                      freq=dtarr.freq, tz=dtarr.tz)
290
291  ->         return subarr._deepcopy_if_needed(ref_to_data, copy)
292
293         @classmethod
294         def _simple_new(cls, values, name=None, freq=None, tz=None, dtype=None):
295             """
296             we require the we have a dtype compat for the values
(Pdb) ref_to_data
<DatetimeArrayMixin>
['2000-01-01 00:00:00', '2000-01-02 00:00:00', '2000-01-03 00:00:00',
 '2000-01-04 00:00:00']
Length: 4, dtype: datetime64[ns]
```

I don't think it's a bug. This is to ensure that `b = pd.DatetimeIndex(a, copy=True)`, don't share a view back to the same ndarray. We need `a._data._data is not b._data._data`.

Reply:
I notice now that this is the same issue we faced in _libs/reduction.pyx, where we need a ref back to the actual concrete array backing the data. In this case we may be able to update Index._deepcopy_if_needed to use the the new Index._index_data attribute, but I don't know how heavily we want to rely on that at this point.

Reply:
@jreback I looked into this a bit more. I think we're able to remove it. It was only ever used in `DatetimeIndex.__new__`

It wasn't hit during our tests, but I don't think there is a case where it was actually needed anymore. sequence_to_dt64ns should take care of copying if needed.

Reply:
Could removing this be separated?